umbral-pre-python/src/lib.rs

```diff
@@ -177,28 +506,50 @@ pub fn decrypt_reencrypted(
     capsule: &Capsule,
     cfrags: Vec<CapsuleFrag>,
     ciphertext: &[u8],
-) -> Option<PyObject> {
+) -> PyResult<PyObject> {
     let backend_cfrags: Vec<umbral_pre::CapsuleFrag> =
         cfrags.iter().cloned().map(|cfrag| cfrag.backend).collect();
-    let res = umbral_pre::decrypt_reencrypted(
+    umbral_pre::decrypt_reencrypted(
         &decrypting_sk.backend,
         &delegating_pk.backend,
         &capsule.backend,
         &backend_cfrags,
         ciphertext,
-    );
-    match res {
-        Some(plaintext) => Some(PyBytes::new(py, &plaintext).into()),
-        None => None,
-    }
+    )
+    .map(|plaintext| PyBytes::new(py, &plaintext).into())
+    .map_err(|err| match err {
+        ReencryptionError::OnOpen(err) => match err {
+            OpenReencryptedError::NoCapsuleFrags => {
+                PyValueError::new_err("Empty CapsuleFrag sequence")
+            }
+            OpenReencryptedError::MismatchedCapsuleFrags => {
+                PyValueError::new_err("CapsuleFrags are not pairwise consistent")
+            }
+            OpenReencryptedError::RepeatingCapsuleFrags => {
+                PyValueError::new_err("Some of the CapsuleFrags are repeated")
+            }
+            // Will be removed when #39 is fixed
+            OpenReencryptedError::ZeroHash => {
+                GenericError::new_err("An internally hashed value is zero")
+            }
+            OpenReencryptedError::ValidationFailed => {
+                GenericError::new_err("Internal validation failed")
+            }
+        },
+        ReencryptionError::OnDecryption(err) => map_decryption_err(err),
+    })
 }
 
 /// A Python module implemented in Rust.
 #[pymodule]
-fn _umbral(_py: Python, m: &PyModule) -> PyResult<()> {
+fn _umbral(py: Python, m: &PyModule) -> PyResult<()> {
     m.add_class::<SecretKey>()?;
+    m.add_class::<SecretKeyFactory>()?;
     m.add_class::<PublicKey>()?;
-    m.add_class::<Parameters>()?;
+    m.add_class::<Capsule>()?;
+    m.add_class::<KeyFrag>()?;
+    m.add_class::<CapsuleFrag>()?;
+    m.add("GenericError", py.get_type::<GenericError>())?;
     m.add_function(wrap_pyfunction!(encrypt, m)?).unwrap();
```

Comment:
Can we also `?` instead of `unwrap()` other calls here?

Reply:
Good point. I took this bit directly from the manual, who know why they put `unwrap()` there.

Reply:
Fixed this in #41